### PR TITLE
Don't alloc during opcode creation

### PR DESCRIPTION
**Pull request by chadhietala**
Probably minimal impact but doesn't degrade code quality a ton. Also replaced some `var` -> `let`.

### CODE REVIEW

File: packages/glimmer-compiler/lib/template-compiler.ts [hunk 1]
```diff
@@ -407,16 +412,16 @@ export default class TemplateCompiler {
...
     if (!pairs.length) {
       this.opcode('pushLiteral', null, null);
       return;
     }
 
-    for (var i = pairs.length - 1; i >= 0; i--) {
-      var key = pairs[i].key;
-      var value = pairs[i].value;
+    for (let i = pairs.length - 1; i >= 0; i--) {
+      let key = pairs[i].key;
```
Comment (stefanpenner) [inline packages/glimmer-compiler/lib/template-compiler.ts]: `let { key, value } = pairs[i];`

Comment (chadhietala) [inline packages/glimmer-compiler/lib/template-compiler.ts]: Fixed.

**Comment by stefanpenner** [comment]
LGTM, mergable when CI is green.

File: packages/glimmer-compiler/lib/template-compiler.ts [hunk 2]
```diff
@@ -264,10 +268,11 @@ export default class TemplateCompiler {
   component([action]) {
     let { attributes, tag } = action;
 
-    attributes.forEach(({ name, value }) => {
+    for (let i = 0, l = attributes.length; i < l; i++) {
```
Comment (stefanpenner) [inline packages/glimmer-compiler/lib/template-compiler.ts]: lets not cache `length`

**Comment by stefanpenner** [comment]
one comment, then it should be good to go.

File: packages/glimmer-compiler/lib/template-compiler.ts [hunk 3]
```diff
@@ -252,9 +252,13 @@ export default class TemplateCompiler {
 
   openElement([action]) {
     this.opcode('openElement', action, action.tag, action.blockParams);
+    for (let i = 0, l = action.attributes.length; i < l; i++) {
+      this.attribute([action.attributes[i]]);
+    }
 
-    action.attributes.forEach(attr => this.attribute([attr]));
-    action.modifiers.forEach(modifier => this.modifier([modifier]));
+    for (let i = 0, l = action.modifiers.length; i < l; i++) {
```
Comment (stefanpenner) [inline packages/glimmer-compiler/lib/template-compiler.ts]: lets not cache length

File: packages/glimmer-compiler/lib/template-compiler.ts [hunk 4]
```diff
@@ -252,9 +252,13 @@ export default class TemplateCompiler {
 
   openElement([action]) {
     this.opcode('openElement', action, action.tag, action.blockParams);
+    for (let i = 0, l = action.attributes.length; i < l; i++) {
```
Comment (stefanpenner) [inline packages/glimmer-compiler/lib/template-compiler.ts]: lets not cache length

**Comment by chadhietala** [comment]
I've been sort of working on a bunch of babel transforms to hopefully make this a non manual endeavor. https://github.com/chadhietala/babel-array-method-zero-cost-abstractions

**Comment by stefanpenner** [comment]
@chadhietala cool, the tricky bit is knowing if the given `forEach` is being called on an array or not.

**Comment by chadhietala** [comment]
:grimacing: yea...

**Comment by stefanpenner** [comment]
@chadhietala a different approach, i took some time ago was to use sweet.js style macro, i think the same approach could work here nicely.

```js
import { forEach } from 'zero-cost-abstractions';

forEach(array, callback) 
```

Then use babel's matcher stuff to detect which forEach it is, now the developer has basically annotated which `forEach`/`map` etc it want's to optimize. such macro fun.

**Comment by chadhietala** [comment]
Yea annotation is what I was thinking

**Comment by chadhietala** [comment]
So on my flight I took a whack at the macros. Was more of experiment in writing sweet.js macros then anything that I would recommending using at this point. https://github.com/chadhietala/array-method-macros

**Pull request merged**